msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/AcquireTokenSilentSupplier.java

```diff
@@ -120,4 +82,74 @@ AuthenticationResult execute() throws Exception {
+            // only because of refreshOn, then simply return the existing cached token
+            if (!afterRefreshOn && (silentRequest.parameters().forceRefresh() || StringHelper.isBlank(cachedResult.accessToken()))) {
+                throw ex;
+            }
+            return cachedResult;
+        }
+    }
+
+    private void setRefreshTelemetry(AuthenticationResult cachedResult) {
+        //As of version 3 of the telemetry schema, there is a field for collecting data about why a token was refreshed,
+        // so here we set the telemetry value based on the cause of the refresh
+        if (silentRequest.parameters().forceRefresh()) {
+            clientApplication.serviceBundle().getServerSideTelemetry().getCurrentRequest().cacheInfo(
+                    CacheTelemetry.REFRESH_FORCE_REFRESH.telemetryValue);
+        } else if (afterRefreshOn) {
+            clientApplication.serviceBundle().getServerSideTelemetry().getCurrentRequest().cacheInfo(
+                    CacheTelemetry.REFRESH_REFRESH_IN.telemetryValue);
+        } else if (cachedResult.expiresOn() < new Date().getTime() / 1000) {
+            clientApplication.serviceBundle().getServerSideTelemetry().getCurrentRequest().cacheInfo(
+                    CacheTelemetry.REFRESH_ACCESS_TOKEN_EXPIRED.telemetryValue);
+        } else if (StringHelper.isBlank(cachedResult.accessToken())) {
+            clientApplication.serviceBundle().getServerSideTelemetry().getCurrentRequest().cacheInfo(
+                    CacheTelemetry.REFRESH_NO_ACCESS_TOKEN.telemetryValue);
+        }
+    }
+
+    //Handles any logic to determine if a token should be refreshed, based on the request parameters and the status of cached tokens
+    private boolean shouldRefresh(SilentParameters parameters, AuthenticationResult cachedResult) {
+
+        //If there is a refresh token but no access token, we should use the refresh token to get the access token
+        if (StringHelper.isBlank(cachedResult.accessToken()) && !StringHelper.isBlank(cachedResult.refreshToken())) {
+            return true;
+        }
+
+        //If the request contains claims then the token should be refreshed, to ensure that the returned token has the correct claims
+        if (parameters.claims() != null) {
+            return true;
+        }
+
+        //Certain long-lived tokens will have a 'refresh on' time that indicates a refresh should be attempted long before the token would expire
+        long currTimeStampSec = new Date().getTime() / 1000;
+        if (cachedResult.refreshOn() != null && cachedResult.refreshOn() > 0 &&
+                cachedResult.refreshOn() < currTimeStampSec && cachedResult.expiresOn() >= currTimeStampSec){
+            afterRefreshOn = true;
+            return true;
+        }
+
+        //Finally, simply return the value of the request's forceRefresh parameter
+        return parameters.forceRefresh();
+    }
 }
```

Comment:
Missing unit tests.

Reply:
Added test in the latest commit specifically for claims being a reason to refresh, other refresh reasons are already covered by existing tests.